### ISSUE TITLE
how to use customer exception handler by APIRouter?

### ISSUE THREAD

**Issue by WhyTimePass** [tiangolo/fastapi#1667]
I want to write different handler with different APIRouter, but when i use APIRouter().exception_handler，the ide notice me no that "AttributeError : 'APIRouter' object has no attribute 'exception_handler'...
  how can i use customer exception handler with different APIRouter?

**Comment by MacMacky** [comment]
@WhyTimePass 
Hi, You can't add an exception handler in an <code>APIRouter</code> instance only on the <code>FastAPI</code> instance.

**Comment by Kludex** [comment]
@WhyTimePass I suggest you to create different errors and handle them in different ways. Example:

```
class PerfectException(HTTPException):
    pass

class TooPerfectException(HTTPException):
    pass

def perfect_exception_handler(request: Request, exc: HTTPException):
    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi"})

def too_perfect_exception_handler(request: Request, exc: HTTPException):
    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi, my friend"})

app = FastAPI()

app.add_exception_handler(PerfectException, perfect_exception_handler)
app.add_exception_handler(TooPerfectExceptionHandler, too_perfect_exception_handler)
```

**Comment by WhyTimePass** [comment]
thanks,

**Comment by Ma233** [comment]
According to the documentation: https://fastapi.tiangolo.com/advanced/custom-request-and-route/
You can also implement it by changing the behaviour of `route_handler`.

**Comment by WhyTimePass** [comment]
> According to the documentation: https://fastapi.tiangolo.com/advanced/custom-request-and-route/
> You can also implement it by changing the behaviour of `route_handler`.

I had a try, but it does not work. ApiRoute instance has not router

**Comment by WhyTimePass** [comment]
> According to the documentation: https://fastapi.tiangolo.com/advanced/custom-request-and-route/
> You can also implement it by changing the behaviour of `route_handler`.

Thank you ! this ways can help me, but it is still hard.
Has there been exist use decorator to handle custom exception

**Comment by michaelkariv** [comment]
I would like to suggest a use case where such functionality is important. I am porting the existing implementation from Google App Engine/Python 2.7. I have errors that are custom. So for some routes 422 error will have a response JSON formatted specifically for that route. So having error handlers be attached to @router, thar then @app would be a huge reprive

**Comment by james77777778** [comment]
I figured out the APIRouter exception handler just like Flask
Hope this helps

(ex: add prometheus_client and logger when HTTPException occurs)
```python
from fastapi import APIRouter, HTTPException, Request, Response
from fastapi.routing import APIRoute
from prometheus_client import Counter

system_fail_count = Counter('system_fail_count', 'Counts the number of system fail')

class PrometheusRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            app = request.app
            try:
                return await original_route_handler(request)
            except HTTPException as exc:
                system_fail_count.inc()
                app.logger.warning('{}'.format(exc.detail))
                raise exc

        return custom_route_handler


router = APIRouter(prefix='/services', route_class=PrometheusRoute)
```

**Comment by wangxin688** [comment]
still want to know if there is a

> @WhyTimePass Hi, You can't add an exception handler in an `APIRouter` instance only on the `FastAPI` instance. Maybe this will change in the future if the maintainer adds this functionality.

Also hope this feature will comming soon. Now, it's difficult to add customize Exception for APIRouter .

**Comment by VikashKothary** [comment]
The reason I wanted to do this is to split the exceptions logic and the main app, so I customised @Kludex's example to support multiple files so it works closer to the `include_router` logic when working with [bigger applications](https://fastapi.tiangolo.com/tutorial/bigger-applications/).

exceptions.py:

```
from fastapi import FastAPI, HTTPException, status
from fastapi.respones import JSONResponse

class PerfectException(HTTPException):
    pass

class TooPerfectException(HTTPException):
    pass

def perfect_exception_handler(request: Request, exc: HTTPException):
    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi"})

def too_perfect_exception_handler(request: Request, exc: HTTPException):
    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi, my friend"})

def include_app(app):
	app.add_exception_handler(PerfectException, perfect_exception_handler)
	app.add_exception_handler(TooPerfectExceptionHandler, too_perfect_exception_handler)
```

app.py

```
import exceptions

app = FastAPI()
exceptions.include_app(app)
```

**Comment by tiangolo** [comment]
Thanks for the discussion everyone! ☕ 

About the original question, I think it's probably a misunderstanding. If you check the docs about custom `APIRoute` class you can see about custom handlers: https://fastapi.tiangolo.com/advanced/custom-request-and-route/#accessing-the-request-body-in-an-exception-handler

But there's not mention of a `exception_handler` attribute in the `APIRouter` class.

Note: have in mind that `APIRoute` and `APIRouter` are two different things.

If anyone else is still having related problems, please create a new issue, with a self-contained example, etc. To be able to discuss each particular case independently. ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue and they piled up through time, but now I'm checking each one in order.

**Comment by lukejs97** [comment]
> The reason I wanted to do this is to split the exceptions logic and the main app, so I customised @Kludex's example to support multiple files so it works closer to the `include_router` logic when working with [bigger applications](https://fastapi.tiangolo.com/tutorial/bigger-applications/#include-the-apirouters-for-users-and-items).
> 
> exceptions.py:
> 
> ```
> from fastapi import FastAPI, HTTPException, status
> from fastapi.respones import JSONResponse
> 
> class PerfectException(HTTPException):
>     pass
> 
> class TooPerfectException(HTTPException):
>     pass
> 
> def perfect_exception_handler(request: Request, exc: HTTPException):
>     return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi"})
> 
> def too_perfect_exception_handler(request: Request, exc: HTTPException):
>     return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi, my friend"})
> 
> def include_app(app):
> 	app.add_exception_handler(PerfectException, perfect_exception_handler)
> 	app.add_exception_handler(TooPerfectExceptionHandler, too_perfect_exception_handler)
> ```
> 
> app.py
> 
> ```
> import exceptions
> 
> app = FastAPI()
> exceptions.include_app(app)
> ```

The problem I'm still having with this, though, is that when I raise an exception in an APIRouter endpoint, it seems to only report the exception on the server side, not on the client side. The client side just receives an Internal Server Error message. Does anyone know how to get this error handling to work with the APIRouter object as well as just the FastAPI object?

There's a similar [issue](https://stackoverflow.com/questions/67008334/raising-exception-from-apirouter-does-not-run-the-exception-handler-returns-500) mentioned on StackOverflow, but the only solution is not accepted and doesn't work for me either.

**Comment by VikashKothary** [comment]
So if you're raising any of the HTTP Exceptions as defined above, then the exception handler should work. But I believe what you want is something to handle an generic Exception. 

I believe I've tried writing an exception handler for it in the same fashion as the HTTP Exceptions but I don't remember if it worked. It would be something like this (note the only difference is the exception type):

```
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware

def server_exception_handler(request: Request, exc: Exception):
    return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": "hi"})

def include_app(app: FastAPI):
	app.add_exception_handler(Exception, server_exception_handler)
```

Alternatively, a workaround that I've tried and I know should work is to use a middleware class to implement a try catch, then return the exception handler. 

```
from fastapi import FastAPI,

async def http_middleware(request: Request, call_next):
    try:
        response = await call_next(request)
        return response
    except Exception as exception:
        return await server_exception_handler(request, exception)


def register_app(app: FastAPI):
    app.add_middleware(BaseHTTPMiddleware, dispatch=http_middleware)

```

I don't know if that answers your question but I hope it helps.

**Comment by lukejs97** [comment]
So you're able to use this with a FastAPI app object that includes an APIRouter object and when you raise a generic Exception inside the APIRouter endpoint function, the custom Exception message is seen by the client that called it? I tried both of your suggestions with that setup and I'm still just getting this message: "Exception: {'custom title': 'custom error message'}" (which is my custom message passed into an Exception) with a stack trace above it and that message is only seen in the server output, not the client output.

**Comment by VikashKothary** [comment]
Yep. The second example with the middleware should hide the stack trace. 

I notice some issues with my example (mainly missing imports) which I've updated but it should work. I would recommend adding some logging to see if the middleware is actually being triggered.

**Comment by ghandic** [comment]
If someone needs an example, I am working on a PR for the APIRouter to handle the errors in a self contained way :) https://github.com/strawberry-graphql/strawberry/pull/2589
